### ISSUE TITLE
Handle overfitting / underfitting issues by adjusting model complexity, regularization techniques

### ISSUE THREAD

**Issue by Sudkart** [Sudkart/DS_ProjectManagement#17]

**Comment by LaxmiKasireddy** [comment]
According to our DNN model ,it seems that the model is performing quite well on the training data, achieving high accuracy and ROC-AUC scores. However, the test accuracy and ROC-AUC scores are slightly lower, which suggests that the model may be slightly overfitting. so we are working on OVERFITTING, considering Regularization, Hyperparameter Tuning and Reducing the complexity of model.

**Comment by LaxmiKasireddy** [comment]
model.fit(X_train_padded, y_train, validation_data=(X_test_padded, y_test), epochs=5, batch_size=32)

Epoch 1/5
254/254 [==============================] - 41s 135ms/step - loss: 0.5279 - accuracy: 0.7308 - val_loss: 0.4381 - val_accuracy: 0.8044
Epoch 2/5
254/254 [==============================] - 32s 125ms/step - loss: 0.3172 - accuracy: 0.8713 - val_loss: 0.4714 - val_accuracy: 0.7867
Epoch 3/5
254/254 [==============================] - 34s 134ms/step - loss: 0.1997 - accuracy: 0.9278 - val_loss: 0.5276 - val_accuracy: 0.7906
Epoch 4/5
254/254 [==============================] - 33s 130ms/step - loss: 0.1376 - accuracy: 0.9512 - val_loss: 0.6209 - val_accuracy: 0.7872
Epoch 5/5
254/254 [==============================] - 32s 125ms/step - loss: 0.1049 - accuracy: 0.9649 - val_loss: 0.6452 - val_accuracy: 0.7842

**Comment by LaxmiKasireddy** [comment]
The training accuracy increases with each epoch, reaching 0.9649 (or approximately 96.49%) at the end of the 5th epoch.

The training loss decreases with each epoch, indicating that the model is learning from the training data.

However, the validation accuracy (val_accuracy) and validation loss (val_loss) show fluctuations and do not improve consistently. The validation accuracy reaches 0.7842 (or approximately 78.42%) at the end of the 5th epoch.

This discrepancy between training accuracy and validation accuracy, along with the fluctuation in validation accuracy, suggests that the model may be overfitting to some extent.

**Comment by LaxmiKasireddy** [comment]
Adam Optimizer
RMSprop Optimizer
Stochastic Gradient Descent (SGD) Optimizer
AdaDelta Optimizer
AdaMax Optimizer
Each optimizer was used to train a model on a specific dataset, and their performance was evaluated in terms of training accuracy, validation accuracy, and area under the ROC curve (AUC).

**Comment by LaxmiKasireddy** [comment]
RMSprop optimizer achieved the highest validation accuracy of approximately 80.89% and the highest AUC of 0.8636 among the listed optimizers. It also showed better training accuracy compared to most of the other optimizers.

The Adam optimizer, with an AUC of 0.8387, and the AdaMax optimizer, with an AUC of 0.8566, also demonstrated good performance, but their validation accuracy was not as high as that of the RMSprop optimizer.

**Comment by LaxmiKasireddy** [comment]
SGD and AdaDelta optimizers performed poorly with relatively low training and validation accuracy, and their AUC values were not significantly better than random guessing.

**Comment by LaxmiKasireddy** [comment]
The best hyperparameters found during the search are:

Dropout rate: 0.3832
Learning rate: 0.0002156
LSTM units: 125.1

During the hyperparameter search, tested various combinations of dropout rate, learning rate, and the number of LSTM units in the model.

The table shows the results of different model configurations during the hyperparameter search. Each row in the table represents a different configuration, with columns indicating the iteration number, the target metric (accuracy in this case), and the values for dropout rate, learning rate, and LSTM units.
